src/components/modebar/attributes.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+var modeBarButtons = require('./buttons');
+var buttonList = Object.keys(modeBarButtons);
+var backButtons = [
+    'v1hovermode',
+    'hoverclosest',
+    'hovercompare',
+    'togglehover',
+    'togglespikelines',
+    'drawclosedpath',
+    'drawopenpath',
+    'drawline',
+    'drawrect',
+    'drawcircle',
+    'eraseshape',
+];
+
+var foreButtons = [];
+var addToForeButtons = function(b) {
+    if(backButtons.indexOf(b._cat || b.name) !== -1) return;
+    // for convenience add lowercase shotname e.g. zoomin as well fullname zoomInGeo
+    var name = b.name;
+    var _cat = (b._cat || b.name).toLowerCase();
+    if(foreButtons.indexOf(name) === -1) foreButtons.push(name);
+    if(foreButtons.indexOf(_cat) === -1) foreButtons.push(_cat);
+};
+buttonList.forEach(function(k) {
+    addToForeButtons(modeBarButtons[k]);
+});
+foreButtons.sort();
+
+module.exports = {
+    editType: 'modebar',
+
+    orientation: {
+        valType: 'enumerated',
+        values: ['v', 'h'],
+        dflt: 'h',
+        editType: 'modebar',
+        description: 'Sets the orientation of the modebar.'
+    },
+    bgcolor: {
+        valType: 'color',
+        editType: 'modebar',
+        description: 'Sets the background color of the modebar.'
+    },
+    color: {
+        valType: 'color',
+        editType: 'modebar',
+        description: 'Sets the color of the icons in the modebar.'
+    },
+    activecolor: {
+        valType: 'color',
+        editType: 'modebar',
+        description: 'Sets the color of the active or hovered on icons in the modebar.'
+    },
+    uirevision: {
+        valType: 'any',
+        editType: 'none',
+        description: [
+            'Controls persistence of user-driven changes related to the modebar,',
+            'including `hovermode`, `dragmode`, and `showspikes` at both the',
+            'root level and inside subplots. Defaults to `layout.uirevision`.'
+        ].join(' ')
+    },
+    add: {
+        valType: 'flaglist',
```

Comment:
Using a flaglist here is convenient for validation, but it's different from the API we use for config, an array. Here we don't need to support the object form, which obviously wouldn't fit in a flaglist, I just wonder if ease of validation is sufficient reason to differ. @nicolaskruchten thoughts?

Reply:
I prefer an array, like in config

Reply:
Addressed in https://github.com/plotly/plotly.js/pull/5660/commits/5c17d6b0899ee22d1af7c634108cabc3aee4dfdd and https://github.com/plotly/plotly.js/pull/5660/commits/329309789b0840c76eaccec3d24a047b92c95078.